File: naive.rb
require 'digest/sha1'
require 'benchmark'

class Array
  def shuffle!
    size.downto(1) { |n| push delete_at(rand(n)) }
    self
  end
end


WORDCHARS = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".split(//)
DICTIONARY = %w{ Cloud Ruby DHH one eight six active record controller data rspec mongrel MySQL postgresSQL tokyo MRI jruby rubinius memcached exception metaprogramming reflection }
PHRASE = "I am not a big believer in fortune telling"

PHRASE_SHA1 = Digest::SHA1.hexdigest(PHRASE)
PHRASE_BINARY = Integer("0b#{PHRASE_SHA1.to_i(base=16).to_s(2)}")

$best_distance = 10000
$best_content = ""

# 10 = 1
# 01 = 1
# 11 = 0
# 00 = 0

def calculate(phrase)
  sha1 = Digest::SHA1.hexdigest(phrase)
  distance = (Integer("0b#{sha1.to_i(base=16).to_s(2)}") ^ PHRASE_BINARY).to_s(2).gsub(/0/, '').length

  return unless distance < $best_distance
  $best_distance = distance
  $best_content = phrase
  puts "#{distance} #{phrase}"
end

def calculate_with_letters(phrase, letters)
  return if letters.length >= 5
  WORDCHARS.each {|l| calculate(phrase + ' ' + letters + l); calculate_with_letters(phrase, letters + l) }
end

def naive
  1.upto(10) do
    phrase = DICTIONARY.shuffle![0,12].join(' ')
    calculate(phrase)
    calculate_with_letters(phrase, "")
  end
end

Benchmark.bm do |x|
  x.report("naive") { naive }
end

File: naive.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "sha1.h"

int WORDCHARS_LENGTH = 62;
char WORDCHARS[] = {"0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"};
char TEST_PHRASE[] = {"Cloud Ruby DHH one eight six active record controller data rspec mongrel\0"};
char PHRASE[] = {"I am not a big believer in fortune telling\0"};

int count = 0;
int phrase_msg[5];

int best_distance = 1000;

SHA1Context sha;

unsigned hamdist(unsigned x, unsigned y)
{
    unsigned dist = 0, val = x ^ y;

    // Count the number of set bits (Knuth's algorithm)
    while(val)
    {
        ++dist; 
        val &= val - 1;
    }

    return dist;
}

void calculate(char *phrase)
{
    SHA1Reset(&sha);
    SHA1Input(&sha, (const unsigned char *) phrase, strlen(phrase));

    if (!SHA1Result(&sha))
    {
        fprintf(stderr, "ERROR-- could not compute message digest\n");
        return;
    }
    else
    {
        int distance = 0;
        for(int i = 0; i < 5 ; i++)
        {
            distance += hamdist(phrase_msg[i], sha.Message_Digest[i]);
        }

        if(distance < best_distance) {
            printf("%d - %s\n", distance, phrase);
            best_distance = distance;
        }
    }
}

void calculate_with_letters(char *phrase, int letter_length)
{
    calculate(phrase);

    count ++;
    if(count > 100000) {
        printf("Done\n");
        exit(0);
    }

    int plength = strlen(phrase);
    int blength;

    if(letter_length == 0) {
        blength = plength + 3;
    } else {
        blength = plength + 2;
    }

    char buffer[blength];
    if(letter_length == 0) {
        sprintf(buffer, "%s ", phrase);
    } else {
        sprintf(buffer, "%s", phrase);
    }

    buffer[blength - 1] = '\0';

    for(int i = 0; i < WORDCHARS_LENGTH; i++) {
        buffer[blength - 2] = WORDCHARS[i];

        calculate(buffer);
        if(letter_length < 4) {
            calculate_with_letters(buffer, letter_length + 1);
        }
    }
}

int main(int argc, char **argv)
{
    SHA1Reset(&sha);
    SHA1Input(&sha, (const unsigned char *) PHRASE, strlen(PHRASE));

    if (!SHA1Result(&sha))
    {
        fprintf(stderr, "ERROR-- could not compute message digest\n");
    }
    else
    {
        for(int i = 0; i < 5 ; i++)
        {
            phrase_msg[i] = sha.Message_Digest[i];
        }
    }

    calculate_with_letters(TEST_PHRASE, 0);
}

